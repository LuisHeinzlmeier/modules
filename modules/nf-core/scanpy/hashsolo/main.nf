process SCANPY_HASHSOLO {
    debug true
    tag "$meta.id"
    label 'process_medium'

    conda "${moduleDir}/environment.yml"
    container "${ workflow.containerEngine == 'singularity' && !task.ext.singularity_pull_docker_container ?
        'oras://community.wave.seqera.io/library/matplotlib_pandas_python_pyyaml_scanpy:e7e659a041ff08e0':
        'community.wave.seqera.io/library/matplotlib_pandas_python_pyyaml_scanpy:bf7b7ef27120d15c' }"

    input:
    val(use_10x)
    tuple val(meta), path(hto_data_10x), path(hto_data_h5ad), val(cell_hashing_columns)

    output:
    tuple val(meta), path("*_hashsolo.csv")       , emit: assignment
    tuple val(meta), path("*_hashsolo.h5ad")      , emit: h5ad
    tuple val(meta), path("*_params_hashsolo.csv"), emit: params
    path "versions.yml", emit: versions

    when:
    task.ext.when == null || task.ext.when

    script:
    prefix = task.ext.prefix ?: "${meta.id}"

    template('hashsolo.py')

    stub:
    prefix = task.ext.prefix ?: "${meta.id}"
    """
    touch ${prefix}_hashsolo.csv
    touch ${prefix}_hashsolo.jpg
    touch ${prefix}_hashsolo.h5ad
    touch ${prefix}_params_hashsolo.csv

    # Prevent failures during scanpy import
    export MPLCONFIGDIR=./tmp/mpl
    export NUMBA_CACHE_DIR=./tmp/numba

    cat <<-END_VERSIONS > versions.yml
    "${task.process}":
        python: \$(python3 --version | cut -f 2 -d " ")
        scanpy: \$(python3 -c "import scanpy; print(scanpy.__version__)")
        matplotlib: \$(python3 -c "import matplotlib; print(matplotlib.__version__)")
        pandas: \$(python3 -c "import pandas; print(pandas.__version__)")
    END_VERSIONS
    """
}
