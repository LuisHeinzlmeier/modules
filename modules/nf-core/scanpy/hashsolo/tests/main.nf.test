nextflow_process {

    name "Test Process SCANPY_HASHSOLO"
    script "../main.nf"
    process "SCANPY_HASHSOLO"

    tag "modules"
    tag "modules_nfcore"
    tag "hashsolo"
    tag "scanpy/hashsolo"
    tag "scanpy"
    tag "untar"

    // ext.args
    def priors                   = [0.20, 0.8, 0.19]
    def pre_existing_clusters    = null
    def clustering_data          = null
    def number_of_noise_barcodes = null
    def round_digits             = 2

    test("hto_data_10x") {

        setup {
            run("UNTAR") {
                script "modules/nf-core/untar/main.nf"
                process {
                """
                input[0] = Channel.of([ [],file(params.modules_testdata_base_path + "/genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/hto.tar.gz", checkIfExists: true)   ])
                """
                }
            }
        }
        when {

            process {
                """
                input[0] = Channel.of(true)
                input[1] = UNTAR.out.untar.map { meta, files ->
                    [ [ id: 'test' ], files, [], [] ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("hto_data_10x - ext.args") {

        config "./nextflow.config"

        setup {
            run("UNTAR") {
                script "modules/nf-core/untar/main.nf"
                process {
                """
                input[0] = Channel.of([ [],file(params.modules_testdata_base_path + "/genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/hto.tar.gz", checkIfExists: true)   ])
                """
                }
            }
        }
        when {

            params{
                module_args = [
                    "--priors",
                    priors.join(" "),
                    pre_existing_clusters ? "--pre_existing_clusters $pre_existing_clusters" : "",
                    clustering_data ? "--clustering_data $clustering_data" : "",
                    number_of_noise_barcodes ? "--number_of_noise_barcodes $number_of_noise_barcodes" : "",
                    round_digits ? "--round_digits $round_digits" : ""
                ].findAll { it != "" }.join(" ")
            }
            process {
                """
                input[0] = Channel.of(true)
                input[1] = UNTAR.out.untar.map { meta, files ->
                    [ [ id: 'test' ], files, [], [] ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }




    test("hto_data_10x - stub") {

        options "-stub"

        setup {
            run("UNTAR") {
                script "modules/nf-core/untar/main.nf"
                process {
                """
                input[0] = Channel.of([ [],file(params.modules_testdata_base_path + "/genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/hto.tar.gz", checkIfExists: true)   ])
                """
                }
            }
        }
        when {

            process {
                """
                input[0] = Channel.of(true)
                input[1] = UNTAR.out.untar.map { meta, files ->
                    [ [ id: 'test' ], files, [], [] ]
                }
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(process.out).match() }
            )
        }
    }

    test("hto_data_h5ad (generated, no real data)") {

        when {
            process {
                """
                input[0] = Channel.of(false)
                input[1] = [
                    [ id: 'test' ],
                    [],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/hashsolo_anndata.h5ad', checkIfExists: true),
                    ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out,
                        process.out.versions.collect { path(it).yaml }
                    ).match() }
            )
        }

    }

    test("hto_data_h5ad (generated, no real data) - ext.args") {

        config "./nextflow.config"

        when {
            params{
                module_args = [
                    "--priors",
                    priors.join(" "),
                    pre_existing_clusters ? "--pre_existing_clusters $pre_existing_clusters" : "",
                    clustering_data ? "--clustering_data $clustering_data" : "",
                    number_of_noise_barcodes ? "--number_of_noise_barcodes $number_of_noise_barcodes" : ""
                ].findAll { it != "" }.join(" ")
            }
            process {
                """
                input[0] = Channel.of(false)
                input[1] = [
                    [ id: 'test' ],
                    [],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/hashsolo_anndata.h5ad', checkIfExists: true),
                    ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out,
                        process.out.versions.collect { path(it).yaml }
                    ).match() }
            )
        }

    }

    test("hto_data_h5ad (generated, no real data) - stub") {

        options "-stub"

        when {
            process {
                """
                input[0] = Channel.of(false)
                input[1] = [
                    [ id: 'test' ],
                    [],
                    file(params.modules_testdata_base_path + 'genomics/homo_sapiens/10xgenomics/cellranger/hashing_demultiplexing/hashsolo_anndata.h5ad', checkIfExists: true),
                    ['0', '1', '2', '3', '4', '5', '6', '7', '8', '9']
                ]
                """
            }
        }

        then {
            assertAll(
                { assert process.success },
                { assert snapshot(
                        process.out,
                        process.out.versions.collect { path(it).yaml }
                    ).match() }
            )
        }

    }
}
