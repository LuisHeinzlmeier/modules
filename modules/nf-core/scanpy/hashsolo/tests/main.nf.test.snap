{
    "hto_data_10x": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,3e7d7ceeb97d93830556311daff388cd",
                            "test_params_hashsolo.csv:md5,12cd33cdc831617c59d5d2ad9a10ad96"
                        ]
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,043358088cc7f369ebf7ac005cca23fe"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,12cd33cdc831617c59d5d2ad9a10ad96"
                    ]
                ],
                "4": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,3e7d7ceeb97d93830556311daff388cd",
                            "test_params_hashsolo.csv:md5,12cd33cdc831617c59d5d2ad9a10ad96"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,043358088cc7f369ebf7ac005cca23fe"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,12cd33cdc831617c59d5d2ad9a10ad96"
                    ]
                ],
                "plot": [
                    
                ],
                "versions": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-09T10:20:23.920872"
    },
    "hto_data_10x - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.jpg:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,8fd954f7fb047e8ad17ff5710299ddf2"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "plot": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.jpg:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8fd954f7fb047e8ad17ff5710299ddf2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-08T21:41:34.938657"
    },
    "hto_data_h5ad (generated, no real data)": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,1c9cd1a41dccbdd67de5dd95636c36dc",
                            "test_params_hashsolo.csv:md5,1f4ac2b59549ef6161f3f63234a7b9f0"
                        ]
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,ea411221d7316441fc413a2b8e7c5749"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,1f4ac2b59549ef6161f3f63234a7b9f0"
                    ]
                ],
                "4": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,1c9cd1a41dccbdd67de5dd95636c36dc",
                            "test_params_hashsolo.csv:md5,1f4ac2b59549ef6161f3f63234a7b9f0"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,ea411221d7316441fc413a2b8e7c5749"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,1f4ac2b59549ef6161f3f63234a7b9f0"
                    ]
                ],
                "plot": [
                    
                ],
                "versions": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ]
            },
            [
                {
                    "SCANPY_HASHSOLO": {
                        "matplotlib": "3.10.5",
                        "pandas": "2.3.1",
                        "python": "3.12.11",
                        "scanpy": "1.11.2"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-09T10:20:52.464803"
    },
    "generated h5ad": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,e1f129270b67e8575dd060d44e1c0c4b"
                    ]
                ],
                "1": [
                    "versions.yml:md5,d4b55c68ad8effa4580cf202d802b007"
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,e1f129270b67e8575dd060d44e1c0c4b"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,d4b55c68ad8effa4580cf202d802b007"
                ]
            },
            [
                {
                    "SCANPY_HASHSOLO": {
                        "python": "3.12.11",
                        "scanpy": "1.11.2"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-02T10:01:12.076734105"
    },
    "hto_data_h5ad (generated, no real data) - ext.args": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,a1c3c62b086c9c790c9addc84aab9fee",
                            "test_params_hashsolo.csv:md5,e6aa54b83efe11f04b57e8cda4d92a5f"
                        ]
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,7764f0ff9d4bfa7f6aae7ba14136bd80"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,e6aa54b83efe11f04b57e8cda4d92a5f"
                    ]
                ],
                "4": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,a1c3c62b086c9c790c9addc84aab9fee",
                            "test_params_hashsolo.csv:md5,e6aa54b83efe11f04b57e8cda4d92a5f"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,7764f0ff9d4bfa7f6aae7ba14136bd80"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,e6aa54b83efe11f04b57e8cda4d92a5f"
                    ]
                ],
                "plot": [
                    
                ],
                "versions": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ]
            },
            [
                {
                    "SCANPY_HASHSOLO": {
                        "matplotlib": "3.10.5",
                        "pandas": "2.3.1",
                        "python": "3.12.11",
                        "scanpy": "1.11.2"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-09T10:33:38.792387"
    },
    "hto_matrix": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,3e7d7ceeb97d93830556311daff388cd",
                            "test_params_hashsolo.csv:md5,c474067bceee23596942159ed2287f41"
                        ]
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,043358088cc7f369ebf7ac005cca23fe"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,c474067bceee23596942159ed2287f41"
                    ]
                ],
                "4": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,3e7d7ceeb97d93830556311daff388cd",
                            "test_params_hashsolo.csv:md5,c474067bceee23596942159ed2287f41"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,043358088cc7f369ebf7ac005cca23fe"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,c474067bceee23596942159ed2287f41"
                    ]
                ],
                "plot": [
                    
                ],
                "versions": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-08T21:32:31.163776"
    },
    "hto_matrix - stub": {
        "content": [
            {
                "0": [
                    [
                        [
                            
                        ],
                        [
                            "[]_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "[]_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    [
                        [
                            
                        ],
                        "[]_hashsolo.jpg:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        [
                            
                        ],
                        "[]_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        [
                            
                        ],
                        "[]_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,8fd954f7fb047e8ad17ff5710299ddf2"
                ],
                "assignment": [
                    [
                        [
                            
                        ],
                        [
                            "[]_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "[]_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        [
                            
                        ],
                        "[]_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "params": [
                    [
                        [
                            
                        ],
                        "[]_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "plot": [
                    [
                        [
                            
                        ],
                        "[]_hashsolo.jpg:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8fd954f7fb047e8ad17ff5710299ddf2"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-07T19:14:54.511322"
    },
    "hto_data_h5ad (generated, no real data) - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "1": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.jpg:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "4": [
                    "versions.yml:md5,8fd954f7fb047e8ad17ff5710299ddf2"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e",
                            "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "plot": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.jpg:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,8fd954f7fb047e8ad17ff5710299ddf2"
                ]
            },
            [
                {
                    "SCANPY_HASHSOLO": {
                        "python": "3.12.11",
                        "scanpy": "1.11.2",
                        "matplotlib": "3.10.5",
                        "pandas": "2.3.1"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-09T09:33:03.99566"
    },
    "generated h5ad - stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,21a56a941355317a8e1ca02d207f64a8"
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,21a56a941355317a8e1ca02d207f64a8"
                ]
            },
            [
                {
                    "SCANPY_HASHSOLO": {
                        "python": "3.12.11",
                        "scanpy": "1.11.2"
                    }
                }
            ]
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-07-02T10:01:21.413803466"
    },
    "hto_data_10x - ext.args": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,a197107805a4c6893ae2019186050476",
                            "test_params_hashsolo.csv:md5,af606f25f5972a278d6f3b64fd03cb9a"
                        ]
                    ]
                ],
                "1": [
                    
                ],
                "2": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,7f93994c6260f51b22ebcbd1229193a9"
                    ]
                ],
                "3": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,af606f25f5972a278d6f3b64fd03cb9a"
                    ]
                ],
                "4": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ],
                "assignment": [
                    [
                        {
                            "id": "test"
                        },
                        [
                            "test_hashsolo.csv:md5,a197107805a4c6893ae2019186050476",
                            "test_params_hashsolo.csv:md5,af606f25f5972a278d6f3b64fd03cb9a"
                        ]
                    ]
                ],
                "h5ad": [
                    [
                        {
                            "id": "test"
                        },
                        "test_hashsolo.h5ad:md5,7f93994c6260f51b22ebcbd1229193a9"
                    ]
                ],
                "params": [
                    [
                        {
                            "id": "test"
                        },
                        "test_params_hashsolo.csv:md5,af606f25f5972a278d6f3b64fd03cb9a"
                    ]
                ],
                "plot": [
                    
                ],
                "versions": [
                    "versions.yml:md5,f6178c11c8b15300ccafa689830c64e3"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.9.2",
            "nextflow": "25.04.3"
        },
        "timestamp": "2025-08-09T10:33:12.417072"
    }
}